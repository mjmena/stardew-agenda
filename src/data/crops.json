[
  {
    "buy": 70,
    "harvest": {},
    "id": "amaranth",
    "name": "Amaranth",
    "sell": 150,
    "growth": 7,
    "start": 56,
    "end": 83
  },
  {
    "buy": 700,
    "harvest": {},
    "id": "ancient_fruit",
    "name": "Ancient Fruit",
    "sell": 550,
    "growth": 28,
    "regrowth": 7,
    "start": 0,
    "end": 83
  },
  {
    "buy": 30,
    "harvest": {},
    "id": "artichoke",
    "name": "Artichoke",
    "sell": 160,
    "growth": 8,
    "start": 56,
    "end": 83
  },
  {
    "buy": 20,
    "harvest": {},
    "id": "beet",
    "name": "Beet",
    "sell": 100,
    "growth": 6,
    "start": 56,
    "end": 83
  },
  {
    "buy": 30,
    "harvest": {},
    "id": "blue_jazz",
    "name": "Blue Jazz",
    "sell": 50,
    "growth": 7,
    "start": 0,
    "end": 27
  },
  {
    "buy": 80,
    "harvest": {
      "extra_chance": 0.02,
      "level_increase": 5,
      "max": 5,
      "min": 3
    },
    "id": "blueberry",
    "name": "Blueberry",
    "sell": 50,
    "growth": 13,
    "regrowth": 4,
    "start": 28,
    "end": 55
  },
  {
    "buy": 50,
    "harvest": {},
    "id": "bok_choy",
    "name": "Bok Choy",
    "sell": 80,
    "growth": 4,
    "start": 56,
    "end": 83
  },
  {
    "buy": 80,
    "harvest": {},
    "id": "cauliflower",
    "name": "Cauliflower",
    "sell": 175,
    "growth": 12,
    "start": 0,
    "end": 27
  },
  {
    "buy": 30,
    "harvest": {
      "extra_chance": 0.02,
      "level_increase": 10,
      "max": 6,
      "min": 4
    },
    "id": "coffee_bean",
    "name": "Coffee Bean",
    "sell": 15,
    "growth": 10,
    "regrowth": 2,
    "start": 0,
    "end": 55
  },
  {
    "buy": 150,
    "harvest": {
      "extra_chance": 0,
      "level_increase": 5,
      "max": 2,
      "min": 1
    },
    "id": "corn",
    "name": "Corn",
    "sell": 50,
    "growth": 14,
    "regrowth": 4,
    "start": 28,
    "end": 83
  },
  {
    "buy": 240,
    "harvest": {
      "extra_chance": 0.1,
      "level_increase": 5,
      "max": 6,
      "min": 2
    },
    "id": "cranberries",
    "name": "Cranberries",
    "sell": 75,
    "growth": 7,
    "regrowth": 5,
    "start": 56,
    "end": 83
  },
  {
    "buy": 20,
    "harvest": {
      "extra_chance": 0.002,
      "level_increase": 10,
      "max": 2,
      "min": 1
    },
    "id": "eggplant",
    "name": "Eggplant",
    "sell": 60,
    "growth": 5,
    "regrowth": 5,
    "start": 56,
    "end": 83
  },
  {
    "buy": 200,
    "harvest": {},
    "id": "fairy_rose",
    "name": "Fairy Rose",
    "sell": 290,
    "growth": 12,
    "start": 56,
    "end": 83
  },
  {
    "buy": 40,
    "harvest": {},
    "id": "garlic",
    "name": "Garlic",
    "sell": 60,
    "growth": 4,
    "start": 0,
    "end": 27
  },
  {
    "buy": 60,
    "harvest": {
      "extra_chance": 0,
      "level_increase": 6,
      "max": 2,
      "min": 1
    },
    "id": "grape",
    "name": "Grape",
    "sell": 80,
    "growth": 10,
    "regrowth": 3,
    "start": 56,
    "end": 83
  },
  {
    "buy": 60,
    "harvest": {
      "extra_chance": 0,
      "level_increase": 6,
      "max": 2,
      "min": 1
    },
    "id": "green_bean",
    "name": "Green Bean",
    "sell": 40,
    "growth": 10,
    "regrowth": 3,
    "start": 0,
    "end": 27
  },
  {
    "buy": 60,
    "harvest": {
      "extra_chance": 0,
      "level_increase": 6,
      "max": 2,
      "min": 1
    },
    "id": "hops",
    "name": "Hops",
    "sell": 25,
    "growth": 11,
    "regrowth": 1,
    "start": 28,
    "end": 55
  },
  {
    "buy": 40,
    "harvest": {
      "extra_chance": 0.03,
      "level_increase": 5,
      "max": 1,
      "min": 1
    },
    "id": "hot_pepper",
    "name": "Hot Pepper",
    "sell": 40,
    "growth": 5,
    "regrowth": 3,
    "start": 28,
    "end": 55
  },
  {
    "buy": 70,
    "harvest": {},
    "id": "kale",
    "name": "Kale",
    "sell": 110,
    "growth": 6,
    "start": 0,
    "end": 27
  },
  {
    "buy": 80,
    "harvest": {},
    "id": "melon",
    "name": "Melon",
    "sell": 250,
    "growth": 12,
    "start": 28,
    "end": 55
  },
  {
    "buy": 20,
    "harvest": {},
    "id": "parsnip",
    "name": "Parsnip",
    "sell": 35,
    "growth": 4,
    "start": 0,
    "end": 27
  },
  {
    "buy": 100,
    "harvest": {},
    "id": "poppy",
    "name": "Poppy",
    "sell": 140,
    "growth": 7,
    "start": 28,
    "end": 55
  },
  {
    "buy": 50,
    "harvest": {
      "extra_chance": 0.2,
      "level_increase": 8,
      "max": 2,
      "min": 1
    },
    "id": "potato",
    "name": "Potato",
    "sell": 80,
    "growth": 6,
    "start": 0,
    "end": 27
  },
  {
    "buy": 100,
    "harvest": {},
    "id": "pumpkin",
    "name": "Pumpkin",
    "sell": 320,
    "growth": 13,
    "start": 56,
    "end": 83
  },
  {
    "buy": 40,
    "harvest": {},
    "id": "radish",
    "name": "Radish",
    "sell": 90,
    "growth": 6,
    "start": 28,
    "end": 55
  },
  {
    "buy": 100,
    "harvest": {},
    "id": "red_cabbage",
    "name": "Red Cabbage",
    "sell": 260,
    "growth": 9,
    "start": 28,
    "end": 55
  },
  {
    "buy": 100,
    "harvest": {},
    "id": "rhubarb",
    "name": "Rhubarb",
    "sell": 220,
    "growth": 13,
    "start": 0,
    "end": 27
  },
  {
    "buy": 400,
    "harvest": {},
    "id": "starfruit",
    "name": "Starfruit",
    "sell": 750,
    "growth": 13,
    "start": 28,
    "end": 55
  },
  {
    "buy": 100,
    "harvest": {
      "extra_chance": 0.02,
      "level_increase": 5,
      "max": 1,
      "min": 1
    },
    "id": "strawberry",
    "name": "Strawberry",
    "sell": 120,
    "growth": 8,
    "regrowth": 4,
    "start": 0,
    "end": 27
  },
  {
    "buy": 50,
    "harvest": {},
    "id": "summer_spangle",
    "name": "Summer Spangle",
    "sell": 90,
    "growth": 8,
    "start": 28,
    "end": 55
  },
  {
    "buy": 125,
    "harvest": {},
    "id": "sunflower",
    "name": "Sunflower",
    "sell": 80,
    "growth": 8,
    "start": 28,
    "end": 83
  },
  {
    "buy": 1000,
    "harvest": {},
    "id": "sweet_gem_berry",
    "name": "Sweet Gem Berry",
    "sell": 3000,
    "growth": 24,
    "start": 56,
    "end": 83
  },
  {
    "buy": 50,
    "harvest": {
      "extra_chance": 0.05,
      "level_increase": 6,
      "max": 1,
      "min": 1
    },
    "id": "tomato",
    "name": "Tomato",
    "sell": 60,
    "growth": 11,
    "regrowth": 4,
    "start": 28,
    "end": 55
  },
  {
    "buy": 20,
    "harvest": {},
    "id": "tulip",
    "name": "Tulip",
    "sell": 30,
    "growth": 6,
    "start": 0,
    "end": 27
  },
  {
    "buy": 10,
    "harvest": {},
    "id": "wheat",
    "name": "Wheat",
    "sell": 25,
    "growth": 4,
    "start": 28,
    "end": 83
  },
  {
    "buy": 60,
    "harvest": {},
    "id": "yam",
    "name": "Yam",
    "sell": 160,
    "growth": 10,
    "start": 56,
    "end": 83
  }
]