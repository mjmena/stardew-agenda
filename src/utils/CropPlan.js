export default class CropPlan {
  constructor({
    date,
    crop,
    fertilizer,
    quantity,
    replant,
    start_date,
    end_date
  }) {
    Object.assign(this, {
      crop,
      fertilizer,
      quantity,
      start_date: date !== undefined ? date : start_date,
      end_date:
        replant || crop.regrowth
          ? crop.end
          : end_date || date + CropPlan.getCropGrowth(crop, fertilizer)
    });
  }

  static getCropGrowth(crop, fertilizer) {
    return crop.growth;
  }

  static merge(planA, planB) {
    if (CropPlan.compare(planA, planB))
      throw new Error(
        "Plans must have equal dates, crop, and fertilizer to merge"
      );
    const { quantity, ...same } = planA;
    return new CropPlan({ quantity: quantity + planB.quantity, ...same });
  }

  // Compares data that cannot be merged - i.e. dates, crop, and fertilizer
  static compare(planA, planB) {
    const start_date_compare = planA.start_date - planB.start_date;
    const crop_compare = planA.crop.id.localeCompare(planB.crop.id);
    const fertilizer_compare = planA.fertilizer.id.localeCompare(
      planB.fertilizer.id
    );
    const end_date_compare = planA.end_date - planB.end_date;

    if (start_date_compare > 0) return 1;
    else if (start_date_compare < 0) return -1;
    else if (crop_compare > 0) return 1;
    else if (crop_compare < 0) return -1;
    else if (fertilizer_compare > 0) return 1;
    else if (fertilizer_compare < 0) return -1;
    else if (end_date_compare > 0) return 1;
    else if (end_date_compare < 0) return -1;
    else return 0;
  }

  static equal(planA, planB) {
    const isSame = CropPlan.compare(planA, planB) === 0;
    return isSame && planA.quantity - planB.quantity === 0;
  }

  shouldReplant = () => {
    return this.date + this.crop.growth === this.end_date;
  };

  get id() {
    return `${this.crop.id}-${this.fertilizer.id}-${this.quantity}-(${
      this.start_date
    }-${this.end_date})`;
  }

  isPlantDate = date => {
    const growth = CropPlan.getCropGrowth(this.crop, this.fertilizer);
    const inPlan = this.start_date <= date && date + growth <= this.end_date;

    return (
      inPlan &&
      (this.crop.regrowth
        ? date === this.start_date
        : (date - this.start_date) % growth === 0)
    );
  };

  isHarvestDate = date => {
    const growth = CropPlan.getCropGrowth(this.crop, this.fertilizer);
    const inPlan = this.start_date <= date - growth && date <= this.end_date;

    return (
      inPlan &&
      (this.crop.regrowth
        ? date === this.start_date + growth ||
          (date - this.start_date - growth) % this.crop.regrowth === 0
        : (date - this.start_date - growth) % growth === 0)
    );
  };

  getAction = type => ({
    id: `${type}-${this.crop.id}`,
    type: type,
    crop: this.crop,
    quantity: this.quantity
  });

  getCropActionsOnDate = date => {
    const isPlant = this.isPlantDate(date);
    const isHarvest = this.isHarvestDate(date);
    if (isPlant && isHarvest)
      return [this.getAction("plant"), this.getAction("harvest")];
    else if (isPlant) return [this.getAction("plant")];
    else if (isHarvest) return [this.getAction("harvest")];
    else return [];
  };
}
