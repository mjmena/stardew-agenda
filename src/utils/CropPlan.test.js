import range from "lodash/range";
import CropPlan from "./CropPlan";
import crops from "../data/crops";
import fertilizers from "../data/fertilizers";

const jazz = crops[4];

const none = fertilizers[0];

const cauliflower = crops[7];
const cauliflower_start_date = 1;
const cauliflower_plan = new CropPlan({
  date: cauliflower_start_date,
  crop: cauliflower,
  fertilizer: none,
  quantity: 1,
  replant: true
});

const jazz_plan = new CropPlan({
  date: 0,
  crop: jazz,
  fertilizer: none,
  quantity: 1
});

const jazz_plan_different_start_date = new CropPlan({
  date: 1,
  crop: jazz,
  fertilizer: none,
  quantity: 1
});

const not_jazz_plan = new CropPlan({
  date: 0,
  crop: cauliflower,
  fertilizer: none,
  quantity: 1
});

const jazz_plan_different_fertilizer = new CropPlan({
  date: 0,
  crop: jazz,
  fertilizer: fertilizers[1],
  quantity: 1
});

const jazz_plan_different_end_date = new CropPlan({
  date: 0,
  crop: jazz,
  fertilizer: none,
  quantity: 1,
  replant: true
});
const jazz_plan_merged = CropPlan.merge(jazz_plan, jazz_plan);

describe("Tests static methods of CropPlan", () => {
  test("Compare a plan to itself", () => {
    expect(CropPlan.compare(jazz_plan, jazz_plan)).toBe(0);
  });

  test("Compare a plan to a similar one (merged)", () => {
    expect(CropPlan.compare(jazz_plan, jazz_plan_merged)).toBe(0);
  });

  test("Compare a plan to a similar one (different start_date only)", () => {
    expect(CropPlan.compare(jazz_plan, jazz_plan_different_start_date)).toBe(
      -1
    );
  });

  test("Compare a plan to a similar one (different crop only)", () => {
    expect(CropPlan.compare(jazz_plan, not_jazz_plan)).toBe(-1);
  });

  test("Compare a plan to a similar one (different fertilizer only)", () => {
    expect(CropPlan.compare(jazz_plan, jazz_plan_different_fertilizer)).toBe(1);
  });

  test("Compare a plan to a similar one (different end_date only)", () => {
    expect(CropPlan.compare(jazz_plan, jazz_plan_different_end_date)).toBe(-1);
  });

  test("Check equality of plans", () => {
    expect(CropPlan.equal(jazz_plan, jazz_plan_merged)).toBe(false);
    expect(CropPlan.equal(jazz_plan, jazz_plan)).toBe(true);
  });
});

describe("Tests for simple CropPlan", () => {
  test("create a CropPlan", () => {
    expect(JSON.stringify(jazz_plan, null, 2)).toMatchSnapshot();
  });
  test("check getCropGrowth", () => {
    expect(CropPlan.getCropGrowth(jazz, none)).toBe(jazz.growth);
  });
  test("Is a plant date?", () => {
    expect(jazz_plan.isPlantDate(0)).toBe(true);
  });

  test("Is not a plant date?", () => {
    expect(jazz_plan.isPlantDate(1)).toBe(false);
  });

  test("Is a harvest date?", () => {
    expect(jazz_plan.isHarvestDate(jazz.growth)).toBe(true);
  });

  test("Is not a harvest date?", () => {
    expect(jazz_plan.isHarvestDate(0)).toBe(false);
  });
  test("get plant action", () => {
    expect(jazz_plan.getCropActionsOnDate(0)).toEqual([
      {
        id: "plant-blue_jazz",
        type: "plant",
        crop: jazz,
        quantity: 1
      }
    ]);
  });
  test("get harvest action", () => {
    expect(jazz_plan.getCropActionsOnDate(jazz.growth)).toEqual([
      {
        id: "harvest-blue_jazz",
        type: "harvest",
        crop: jazz,
        quantity: 1
      }
    ]);
  });
  test("get actions for month", () => {
    const actions = range(0, 28).map(jazz_plan.getCropActionsOnDate);
    expect(actions).toMatchSnapshot();
  });
});

describe("Tests for CropPlan with replant", () => {
  test("create a CropPlan", () => {
    expect(JSON.stringify(cauliflower_plan, null, 2)).toMatchSnapshot();
  });

  test("check getCropGrowth", () => {
    expect(CropPlan.getCropGrowth(cauliflower, none)).toBe(cauliflower.growth);
  });

  test("Is a plant date?", () => {
    expect(cauliflower_plan.isPlantDate(cauliflower_start_date)).toBe(true);
  });

  test("Is not a plant date?", () => {
    expect(cauliflower_plan.isPlantDate(cauliflower_start_date + 1)).toBe(
      false
    );
  });

  test("Is a harvest date?", () => {
    expect(
      cauliflower_plan.isHarvestDate(
        cauliflower_start_date + cauliflower.growth
      )
    ).toBe(true);
  });

  test("Is not a harvest date?", () => {
    expect(cauliflower_plan.isHarvestDate(0)).toBe(false);
  });

  test("get plant action", () => {
    expect(
      cauliflower_plan.getCropActionsOnDate(cauliflower_start_date)
    ).toMatchSnapshot();
  });

  test("get plant and harvest action", () => {
    expect(
      cauliflower_plan.getCropActionsOnDate(
        cauliflower_start_date + cauliflower.growth
      )
    ).toMatchSnapshot();
  });

  test("get harvest action", () => {
    expect(
      cauliflower_plan.getCropActionsOnDate(
        cauliflower_start_date + cauliflower.growth
      )
    ).toMatchSnapshot();
  });

  test("get actions for month", () => {
    expect(
      range(0, 28).map(cauliflower_plan.getCropActionsOnDate)
    ).toMatchSnapshot();
  });
});

const fruit = crops[1];
const fruit_start_date = 0;
const fruit_plan = new CropPlan({
  date: fruit_start_date,
  crop: fruit,
  fertilizer: none,
  quantity: 1,
  replant: true
});

describe("Tests for CropPlan with regrowth", () => {
  test("create a CropPlan", () => {
    expect(JSON.stringify(fruit_plan, null, 2)).toMatchSnapshot();
  });

  test("check getCropGrowth", () => {
    expect(CropPlan.getCropGrowth(fruit, none)).toBe(fruit.growth);
  });

  test("Is a plant date?", () => {
    expect(fruit_plan.isPlantDate(fruit_start_date)).toBe(true);
  });

  test("Is not a plant date?", () => {
    expect(fruit_plan.isPlantDate(fruit_start_date + 1)).toBe(false);
  });

  test("Is a harvest date?", () => {
    expect(fruit_plan.isHarvestDate(fruit_start_date + fruit.growth)).toBe(
      true
    );
  });

  test("Is another harvest date?", () => {
    expect(
      fruit_plan.isHarvestDate(fruit_start_date + fruit.growth + fruit.regrowth)
    ).toBe(true);
  });

  test("Is not a harvest date?", () => {
    expect(fruit_plan.isHarvestDate(0)).toBe(false);
  });

  test("get plant action", () => {
    expect(fruit_plan.getCropActionsOnDate(fruit_start_date)).toMatchSnapshot();
  });

  test("get harvest action", () => {
    expect(
      fruit_plan.getCropActionsOnDate(fruit_start_date + fruit.growth)
    ).toMatchSnapshot();
  });

  test("get actions for year", () => {
    const actions = range(0, 127).map(fruit_plan.getCropActionsOnDate);
    expect(actions).toMatchSnapshot();
  });
});
