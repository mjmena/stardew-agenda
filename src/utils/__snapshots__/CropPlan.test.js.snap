// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for CropPlan with regrowth create a CropPlan 1`] = `
"{
  \\"crop\\": {
    \\"buy\\": 700,
    \\"harvest\\": {},
    \\"id\\": \\"ancient_fruit\\",
    \\"name\\": \\"Ancient Fruit\\",
    \\"sell\\": 550,
    \\"growth\\": 28,
    \\"regrowth\\": 7,
    \\"start\\": 0,
    \\"end\\": 83
  },
  \\"fertilizer\\": {
    \\"buy\\": 0,
    \\"id\\": \\"none\\",
    \\"name\\": \\"None\\"
  },
  \\"quantity\\": 1,
  \\"start_date\\": 0,
  \\"end_date\\": 83
}"
`;

exports[`Tests for CropPlan with regrowth get actions for year 1`] = `
Array [
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "plant-ancient_fruit",
      "quantity": 1,
      "type": "plant",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 700,
        "end": 83,
        "growth": 28,
        "harvest": Object {},
        "id": "ancient_fruit",
        "name": "Ancient Fruit",
        "regrowth": 7,
        "sell": 550,
        "start": 0,
      },
      "id": "harvest-ancient_fruit",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`Tests for CropPlan with regrowth get harvest action 1`] = `
Array [
  Object {
    "crop": Object {
      "buy": 700,
      "end": 83,
      "growth": 28,
      "harvest": Object {},
      "id": "ancient_fruit",
      "name": "Ancient Fruit",
      "regrowth": 7,
      "sell": 550,
      "start": 0,
    },
    "id": "harvest-ancient_fruit",
    "quantity": 1,
    "type": "harvest",
  },
]
`;

exports[`Tests for CropPlan with regrowth get plant action 1`] = `
Array [
  Object {
    "crop": Object {
      "buy": 700,
      "end": 83,
      "growth": 28,
      "harvest": Object {},
      "id": "ancient_fruit",
      "name": "Ancient Fruit",
      "regrowth": 7,
      "sell": 550,
      "start": 0,
    },
    "id": "plant-ancient_fruit",
    "quantity": 1,
    "type": "plant",
  },
]
`;

exports[`Tests for CropPlan with replant create a CropPlan 1`] = `
"{
  \\"crop\\": {
    \\"buy\\": 80,
    \\"harvest\\": {},
    \\"id\\": \\"cauliflower\\",
    \\"name\\": \\"Cauliflower\\",
    \\"sell\\": 175,
    \\"growth\\": 12,
    \\"start\\": 0,
    \\"end\\": 27
  },
  \\"fertilizer\\": {
    \\"buy\\": 0,
    \\"id\\": \\"none\\",
    \\"name\\": \\"None\\"
  },
  \\"quantity\\": 1,
  \\"start_date\\": 1,
  \\"end_date\\": 27
}"
`;

exports[`Tests for CropPlan with replant get actions for month 1`] = `
Array [
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 80,
        "end": 27,
        "growth": 12,
        "harvest": Object {},
        "id": "cauliflower",
        "name": "Cauliflower",
        "sell": 175,
        "start": 0,
      },
      "id": "plant-cauliflower",
      "quantity": 1,
      "type": "plant",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 80,
        "end": 27,
        "growth": 12,
        "harvest": Object {},
        "id": "cauliflower",
        "name": "Cauliflower",
        "sell": 175,
        "start": 0,
      },
      "id": "plant-cauliflower",
      "quantity": 1,
      "type": "plant",
    },
    Object {
      "crop": Object {
        "buy": 80,
        "end": 27,
        "growth": 12,
        "harvest": Object {},
        "id": "cauliflower",
        "name": "Cauliflower",
        "sell": 175,
        "start": 0,
      },
      "id": "harvest-cauliflower",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 80,
        "end": 27,
        "growth": 12,
        "harvest": Object {},
        "id": "cauliflower",
        "name": "Cauliflower",
        "sell": 175,
        "start": 0,
      },
      "id": "harvest-cauliflower",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
]
`;

exports[`Tests for CropPlan with replant get harvest action 1`] = `
Array [
  Object {
    "crop": Object {
      "buy": 80,
      "end": 27,
      "growth": 12,
      "harvest": Object {},
      "id": "cauliflower",
      "name": "Cauliflower",
      "sell": 175,
      "start": 0,
    },
    "id": "plant-cauliflower",
    "quantity": 1,
    "type": "plant",
  },
  Object {
    "crop": Object {
      "buy": 80,
      "end": 27,
      "growth": 12,
      "harvest": Object {},
      "id": "cauliflower",
      "name": "Cauliflower",
      "sell": 175,
      "start": 0,
    },
    "id": "harvest-cauliflower",
    "quantity": 1,
    "type": "harvest",
  },
]
`;

exports[`Tests for CropPlan with replant get plant action 1`] = `
Array [
  Object {
    "crop": Object {
      "buy": 80,
      "end": 27,
      "growth": 12,
      "harvest": Object {},
      "id": "cauliflower",
      "name": "Cauliflower",
      "sell": 175,
      "start": 0,
    },
    "id": "plant-cauliflower",
    "quantity": 1,
    "type": "plant",
  },
]
`;

exports[`Tests for CropPlan with replant get plant and harvest action 1`] = `
Array [
  Object {
    "crop": Object {
      "buy": 80,
      "end": 27,
      "growth": 12,
      "harvest": Object {},
      "id": "cauliflower",
      "name": "Cauliflower",
      "sell": 175,
      "start": 0,
    },
    "id": "plant-cauliflower",
    "quantity": 1,
    "type": "plant",
  },
  Object {
    "crop": Object {
      "buy": 80,
      "end": 27,
      "growth": 12,
      "harvest": Object {},
      "id": "cauliflower",
      "name": "Cauliflower",
      "sell": 175,
      "start": 0,
    },
    "id": "harvest-cauliflower",
    "quantity": 1,
    "type": "harvest",
  },
]
`;

exports[`Tests for simple CropPlan create a CropPlan 1`] = `
"{
  \\"crop\\": {
    \\"buy\\": 30,
    \\"harvest\\": {},
    \\"id\\": \\"blue_jazz\\",
    \\"name\\": \\"Blue Jazz\\",
    \\"sell\\": 50,
    \\"growth\\": 7,
    \\"start\\": 0,
    \\"end\\": 27
  },
  \\"fertilizer\\": {
    \\"buy\\": 0,
    \\"id\\": \\"none\\",
    \\"name\\": \\"None\\"
  },
  \\"quantity\\": 1,
  \\"start_date\\": 0,
  \\"end_date\\": 7
}"
`;

exports[`Tests for simple CropPlan get actions for month 1`] = `
Array [
  Array [
    Object {
      "crop": Object {
        "buy": 30,
        "end": 27,
        "growth": 7,
        "harvest": Object {},
        "id": "blue_jazz",
        "name": "Blue Jazz",
        "sell": 50,
        "start": 0,
      },
      "id": "plant-blue_jazz",
      "quantity": 1,
      "type": "plant",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "crop": Object {
        "buy": 30,
        "end": 27,
        "growth": 7,
        "harvest": Object {},
        "id": "blue_jazz",
        "name": "Blue Jazz",
        "sell": 50,
        "start": 0,
      },
      "id": "harvest-blue_jazz",
      "quantity": 1,
      "type": "harvest",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;
